src/meta/src/barrier/creating_job/mod.rs

```diff
@@ -182,52 +158,22 @@ impl CreatingStreamingJobControl {
     }
 
     pub(super) fn pinned_upstream_log_epoch(&self) -> Option<u64> {
-        let stop_consume_log_store_epoch = match &self.status {
-            CreatingStreamingJobStatus::ConsumingSnapshot { .. }
-            | CreatingStreamingJobStatus::ConsumingLogStore { .. } => None,
-            CreatingStreamingJobStatus::ConsumingUpstream {
-                start_consume_upstream_epoch,
-                ..
-            }
-            | CreatingStreamingJobStatus::Finishing {
-                start_consume_upstream_epoch,
-                ..
-            } => Some(*start_consume_upstream_epoch),
-        };
-        if let Some(max_collected_epoch) = self.barrier_control.max_collected_epoch() {
-            if max_collected_epoch < self.backfill_epoch {
-                Some(self.backfill_epoch)
-            } else if let Some(stop_consume_log_store_epoch) = stop_consume_log_store_epoch
-                && max_collected_epoch >= stop_consume_log_store_epoch
-            {
-                None
-            } else {
-                Some(max_collected_epoch)
-            }
-        } else {
-            Some(self.backfill_epoch)
+        if matches!(&self.status, CreatingStreamingJobStatus::Finishing(_)) {
+            return None;
         }
+        // TODO: when supporting recoverable snapshot backfill, we should use the max epoch that has committed
+        Some(max(
+            self.barrier_control.max_collected_epoch().unwrap_or(0),
```

Comment:
It's the max between backfill_epoch and max_collected_epoch because of two scenarios:
1. If max_collected_epoch is larger than backfill epoch, means that backfill has already started consuming logstore, and has at least finished reading logstore from max_collected_epoch.
2. If backfill epoch is larger, means that backfill has not started logstore read. We need to pin the backfill epoch, as that's where it will start logstore read from.

Reply:
Is my understanding correct?

Reply:
Yes, it's correct.